bindings/rust/evmc-client/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+/* EVMC: Ethereum Client-VM Connector API.
+ * Copyright 2019 The EVMC Authors.
+ * Licensed under the Apache License, Version 2.0.
+ */
+
+pub mod host;
+mod loader;
+pub mod types;
+pub use self::loader::EvmcLoaderErrorCode;
+
+use crate::loader::evmc_load_and_create;
+use crate::types::*;
+use evmc_sys as ffi;
+use std::ffi::CStr;
+
+extern "C" {
+    fn evmc_create() -> *mut ffi::evmc_vm;
+}
+
+fn error(err: EvmcLoaderErrorCode) -> Result<EvmcLoaderErrorCode, &'static str> {
+    match err {
+        EvmcLoaderErrorCode::EvmcLoaderSucces => Ok(EvmcLoaderErrorCode::EvmcLoaderSucces),
+        EvmcLoaderErrorCode::EvmcLoaderCannotOpen => Err("evmc loader: library cannot open"),
+        EvmcLoaderErrorCode::EvmcLoaderSymbolNotFound => {
+            Err("evmc loader: the EVMC create function not found")
+        }
+        EvmcLoaderErrorCode::EvmcLoaderInvalidArgument => {
+            panic!("evmc loader: filename argument is invalid")
+        }
+        EvmcLoaderErrorCode::EvmcLoaderInstanceCreationFailure => {
+            Err("evmc loader: VM instance creation failure")
+        }
+        EvmcLoaderErrorCode::EvmcLoaderAbiVersionMismatch => {
+            Err("evmc loader: ABI version mismatch")
+        }
+        _ => Err("evmc loader: unexpected error"),
+    }
+}
+
+pub struct EvmcVm {
+    handle: *mut ffi::evmc_vm,
+    host_context: *mut ffi::evmc_host_context,
+    host_interface: *mut ffi::evmc_host_interface,
+}
+
+impl EvmcVm {
+    pub fn get_abi_version(&self) -> i32 {
+        unsafe {
+            let version: i32 = (*self.handle).abi_version;
+            version
+        }
+    }
+
+    pub fn get_name(&self) -> &str {
+        unsafe {
+            let c_str: &CStr = CStr::from_ptr((*self.handle).name);
+            c_str.to_str().unwrap()
+        }
+    }
+
+    pub fn get_version(&self) -> &str {
+        unsafe {
+            let c_str: &CStr = CStr::from_ptr((*self.handle).version);
+            c_str.to_str().unwrap()
+        }
+    }
+
+    pub fn destroy(&self) {
+        unsafe { ((*self.handle).destroy.unwrap())(self.handle) }
+    }
+
+    pub fn execute(
+        &self,
+        host_interface: Box<dyn host::HostInterface>,
+        rev: Revision,
+        kind: CallKind,
+        is_static: bool,
+        depth: i32,
+        gas: i64,
+        destination: &Address,
+        sender: &Address,
+        input: &Bytes,
+        value: &Bytes32,
+        code: &Bytes,
+        create2_salt: &Bytes32,
+    ) -> (&Bytes, i64, StatusCode) {
```

Comment:
It would be nice to use `evmc-vm::ExecutionResult` here.

Reply:
The struct is incompatible with my idea. I try to hide ffi layer's nest struct to user make it more user friendly.
ex. in bindings.rs
```rust
pub struct evmc_address {
    #[doc = " The 20 bytes of the hash."]
    pub bytes: [u8; 20usize],
}
```
evmc-vm expose those nest structures directly like Address, Bytes32 ...etc and ExecutionResult include one Address field.
But I usually prefer use the simply types that re-defined in emc-client as below.
```rust
pub const ADDRESS_LENGTH: usize = 20;
pub const BYTES32_LENGTH: usize = 32;
pub type Address = [u8; ADDRESS_LENGTH];
pub type Bytes32 = [u8; BYTES32_LENGTH];
```
I need confirm which way is better.

Reply:
Sorry for the delay, but `evmc-vm::ExecutionResult` is not ffi, it has high-level types.

Reply:
Also regarding the type aliases, the goal was to eventually replace them with nicer types, but to do it incrementally.